### ISSUE TITLE
turn `enum` into `@enum`

### ISSUE THREAD

**Issue by Maiori44** [ClueLang/Clue#85]
when converting old 2.5 code to 3.0 I noticed I changed all enums to a bunch of `@define`, which made me realise `enum` would work better as a preprocessor directive that sets many preprocessor variables

**Comment by Markos-Th09** [comment]
> when converting old 2.5 code to 3.0 I noticed I changed all enums to a bunch of `@define`, which made me realise `enum` would work better as a preprocessor directive that sets many preprocessor variables

The current implementation of enums is kinda bad already because it doesn't really have named enums (which I prefer).

**Comment by Maiori44** [comment]
If you mean Rust's those are too complex and don't feel necessary in Clue (since table)

**Comment by Markos-Th09** [comment]
> If you mean Rust's those are too complex and don't feel necessary in Clue (since tables)

No that. I just mean that the enum should have a name. The variants won't really have any value with them. 

```rust
enum Something{
     A,
     B
}
```

**Comment by Maiori44** [comment]
A name would really only be useful for type checking, which is not yet a thing

**Comment by Markos-Th09** [comment]
> A name would really only be useful for type checking, which is not yet a thing

but we don't really need types to implement it it would just be compiled to a table

**Comment by Maiori44** [comment]
The point of enums is to define state constants more quickly
But since those states are just constant numbers it would be better to make them a compile time thing

**Comment by Markos-Th09** [comment]
> The point of enums is to define state constants more quickly
> But since those states are just constant numbers it would be better to make them a compile time thing

but we still need good lua interoperability

**Comment by Maiori44** [comment]
What?

**Comment by Markos-Th09** [comment]
if you were to make a library which should also work in lua the the enum must be easy to use to use in lua

**Comment by Maiori44** [comment]
I guess that's fair

**Comment by Maiori44** [comment]
How about just having both enum and @enum?
One compiles to global variables that can be used outside of the Clue code (for libraries) and the other is all compile time?

**Comment by Markos-Th09** [comment]
I guess if the difference is clear for everyone it they can satisfy both use cases

**Comment by Maiori44** [comment]
`local enum` could error saying to use `@enum` instead
While `global enum` will work but give a warning

**Comment by Markos-Th09** [comment]
> `local enum` could error saying to use `@enum` instead
> While `global enum` will work but give a warning

no need

**Comment by Maiori44** [comment]
I mean `local enum` would be deprecated so it would error anyway

**Comment by Markos-Th09** [comment]
> I mean `local enum` would be deprecated so it would error anyway

what if somebody wanted to just use a named enum as a local variable?

**Comment by Maiori44** [comment]
They use `@enum`

**Comment by Markos-Th09** [comment]
> They use `@enum`

but @enum is not a named enum, which has some use cases outside of a library

**Comment by Maiori44** [comment]
I doubt naming enums will be useful..

**Comment by Markos-Th09** [comment]
> I doubt naming enums will be useful..

it still has its uses and I would use it for some things

**Comment by Maiori44** [comment]
Forgot to reply-
Ill need to reflect on this

**Comment by ashifolfi** [comment]
> I doubt naming enums will be useful..

I use them all the time they are extremely useful

**Comment by Maiori44** [comment]
You use the _name_ all the time?

**Comment by ashifolfi** [comment]
> You use the _name_ all the time?

Yes I name my enums. Being able to have a single named enum with values of simple names like say FileMode.Read (C# style enum) is very very helpful for code readability. Even in my own Lua enum implementation I use a named enum setup.

I see enums as a group of number constants that belong to a single "type" which is the name.

**Comment by ashifolfi** [comment]
Enums are almost always prefixed with a shared name for organization so why not just group them under that name with a named enum system.

**Comment by Maiori44** [comment]
Because Clue does not have types (yet) so the name would not be used

**Comment by ashifolfi** [comment]
> Because Clue does not have types (yet) so the name would not be used

?????? That doesn't mean anything? They should all have the type be used somewhere in calling them? Like C# does Type.Value and my Lua implementation initially used Type_Value.

**Comment by ashifolfi** [comment]
The point is making them visibly related to eachother along with type checking. Not exclusively type checking.

**Comment by Maiori44** [comment]
How could Clue store a type inside a number?

**Comment by ashifolfi** [comment]
> How could Clue store a type inside a number?

By defining the type as the name of an enum you'd tell the compiler "hey if the provided value doesn't come from an enum value in this name group throw an error"
Anything else with compiling down to Lua could be done however you see fit but the compiler would just check the provided enum value against those present in the enum name you've specified as the type

**Comment by Maiori44** [comment]
Could I have an example of valid and invalid code?
It's kinda late so I can't visualize it well

**Comment by ashifolfi** [comment]
> Could I have an example of valid and invalid code? It's kinda late so I can't visualize it well

```
enum Enum1 {
    Yes,
    No
}

enum Enum2 {
    Yes,
    No
}

Enum1 thing = Enum1.yes // valid as the name comparison checks out and yes is a value in Enum1

Enum1 thing = Enum2.yes // invalid as the value comes from Enum2

Enum1 thing = Enum1.maybe // invalid as maybe is not a value in Enum1

Enum1 thing = Enum3.yes // invalid as Enum3 does not exist
```

Here's a quick example

**Comment by Maiori44** [comment]
I see, makes more sense now
This could likely be done in 4.0, alongside all the other type things
